### PR TITLE
detect/dataset: delay set operation after signature full match

### PR DESCRIPTION
**Pull request by catenacyber**
Link to ticket: https://redmine.openinfosecfoundation.org/issues/
https://redmine.openinfosecfoundation.org/issues/5576

Describe changes:
- detect/dataset: delay set operation after signature full match

SV_BRANCH=https://github.com/OISF/suricata-verify/pull/2000

https://github.com/OISF/suricata/pull/11704 with clean CI

No longer the limitation described for flowvar in https://redmine.openinfosecfoundation.org/issues/7197

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 

ERROR: QA failed on SURI_TLPR1_alerts_cmp.
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPW2_autofp_stats_chk | | |
| .flow.end.tcp_state.established | 8 | 9 | 112.5% | 

Pipeline 22368

File: src/detect-dataset.c [hunk 1]
```diff
@@ -34,25 +34,94 @@
...
+                }
+            } else if (a->v2.Callback == DetectEngineInspectMultiBufferGeneric) {
+                uint32_t local_id = 0;
+                while (1) {
+                    InspectionBuffer *buffer = a->v2.GetMultiData(
+                            det_ctx, a->v2.transforms, f, flags, txv, sd->list, local_id);
+                    if (buffer == NULL || buffer->inspect == NULL)
+                        break;
+                    KEYWORD_PROFILING_PAUSE;
+                    const bool match =
```
Comment (victorjulien) [inline src/detect-dataset.c:73]: this reruns inspection?

File: src/detect-dataset.c [hunk 2]
```diff
@@ -34,25 +34,94 @@
...
 void DetectDatasetFree (DetectEngineCtx *, void *);
 
+static int DetectDatasetTxMatch(DetectEngineThreadCtx *det_ctx, Flow *f, uint8_t flags, void *state,
+        void *txv, const Signature *s, const SigMatchCtx *ctx)
+{
+    const DetectDatasetData *sd = (DetectDatasetData *)ctx;
+    // This is only run for DETECT_SM_LIST_POSTMATCH
+    DEBUG_VALIDATE_BUG_ON(sd->cmd != DETECT_DATASET_CMD_SET);
+
+    DetectEngineAppInspectionEngine *a = s->app_inspect;
```
Comment (victorjulien) [inline src/detect-dataset.c:56]: are we looping this to find our inspect engine? Can't we hard code that instead?

**Comment by victorjulien** [comment]
needs a lot more explanation of how it works and why, as a lot is going on

File: src/detect-dataset.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-dataset.c:56]: > are we looping this to find our inspect engine?

yes

> Can't we hard code that instead?

Yes please, how ?

Here I use `list = s->init_data->list;` from `DetectDatasetSetup`

File: src/detect-dataset.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-dataset.c:73]: yes.

What do you expect from the rule ?
`alert dns any any -> any any (sid: 1; dns.query; content: "suricata.io"; dataset: set,suricataiodomains;)`
When a DNS packet has 2 queries, one for ctf.suricata.io and the other for suricon.net ?

(The alternative to rerunning this detections seems to me to store something, but we do not have a good storage...)

**Comment by catenacyber** [comment]
> needs a lot more explanation of how it works and why, as a lot is going on

Yes, will do...

Feedback is already welcome if you have more :-)

**Pull request closed without merging**